File: part1-functional/solution.js
/**
 * Remove duplicates from an array while preserving the order
 * of first occurrences.
 *
 * @param {Array} arr - The input array containing elements.
 * @returns {Array} A new array with duplicate elements removed.
 */
function removeDuplicates(arr) {
  const seenSet = new Set(); 

  const isNewElement = (x) => {
    if (!seenSet.has(x)) {
      seenSet.add(x);
      return true;
    }
    return false;
  };

  return arr.filter(isNewElement);
}

// Printing test cases below for reviewer convenience — can be disregarded if unnecessary.
console.log(removeDuplicates([1, 2, 3, 2, 4, 1, 5])); // [1, 2, 3, 4, 5]
console.log(removeDuplicates([1, 1, 1]));             // [1]
console.log(removeDuplicates([]));                    // []

File: part1-functional/solution.py
def remove_duplicates(nums):
    """
    Remove duplicates from a list while preserving the order of first occurrences.
    
    Args:
        nums (list): The input list containing elements.
    
    Returns:
        list: A new list with duplicates removed in original order.
    """
    seen_set = set()

    def is_new_num(num):
        if num not in seen_set:
            seen_set.add(num)
            return True
        return False
    return list(filter(is_new_num, nums))

# Printing test cases below for reviewer convenience — can be disregarded if unnecessary.
print(remove_duplicates([1, 2, 3, 2, 4, 1, 5]))  # [1, 2, 3, 4, 5]
print(remove_duplicates([1, 1, 1]))              # [1]
print(remove_duplicates([]))                     # []